File: src/assembler/cicd-assembler.ts
/**
 * CicdAssembler — generates CI/CD artifacts conditionally.
 *
 * Produces: GitHub Actions CI workflow, Dockerfile, Docker Compose,
 * Kubernetes manifests, smoke test config, and deploy runbook.
 * Conditional logic based on container, orchestrator, and smokeTests.
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import type { AssembleResult } from "./rules-assembler.js";
import {
  LANGUAGE_COMMANDS,
  FRAMEWORK_PORTS,
  FRAMEWORK_HEALTH_PATHS,
  DOCKER_BASE_IMAGES,
  DEFAULT_PORT_FALLBACK,
  DEFAULT_HEALTH_PATH,
  DEFAULT_DOCKER_IMAGE,
} from "../domain/stack-mapping.js";

const CICD_TEMPLATES = "cicd-templates";
const CI_TEMPLATE = "ci-workflow/ci.yml.njk";
const COMPOSE_TEMPLATE = "docker-compose/docker-compose.yml.njk";
const RUNBOOK_TEMPLATE = "deploy-runbook/deploy-runbook.md.njk";
const SMOKE_SOURCE = "smoke-tests/smoke-config.md";
const DOCKER_CONDITION = "docker";
const K8S_CONDITION = "kubernetes";

const K8S_MANIFESTS: readonly string[] = [
  "deployment.yaml",
  "service.yaml",
  "configmap.yaml",
];

/** Build stack-specific template context from mapping constants. */
function buildStackContext(
  config: ProjectConfig,
): Record<string, unknown> {
  const langKey = `${config.language.name}-${config.framework.buildTool}`;
  const commands = LANGUAGE_COMMANDS[langKey];
  const port = FRAMEWORK_PORTS[config.framework.name]
    ?? DEFAULT_PORT_FALLBACK;
  const healthPath = FRAMEWORK_HEALTH_PATHS[config.framework.name]
    ?? DEFAULT_HEALTH_PATH;
  const baseImage = DOCKER_BASE_IMAGES[config.language.name]
    ?? DEFAULT_DOCKER_IMAGE;
  const resolvedImage = baseImage.replace(
    "{version}", config.language.version,
  );
  return {
    compile_cmd: commands?.compileCmd ?? "",
    build_cmd: commands?.buildCmd ?? "",
    test_cmd: commands?.testCmd ?? "",
    coverage_cmd: commands?.coverageCmd ?? "",
    file_extension: commands?.fileExtension ?? "",
    build_file: commands?.buildFile ?? "",
    package_manager: commands?.packageManager ?? "",
    framework_port: port,
    health_path: healthPath,
    docker_base_image: resolvedImage,
  };
}

/** Renders a template and writes result to disk. */
function renderAndWrite(
  engine: TemplateEngine,
  templateRelPath: string,
  destPath: string,
  extraContext: Record<string, unknown>,
): boolean {
  try {
    const content = engine.renderTemplate(
      `${CICD_TEMPLATES}/${templateRelPath}`,
      extraContext,
    );
    fs.mkdirSync(path.dirname(destPath), { recursive: true });
    fs.writeFileSync(destPath, content, "utf-8");
    return true;
  } catch {
    return false;
  }
}

/** Generates CI/CD artifacts conditionally based on config. */
export class CicdAssembler {
  /** Assemble CI/CD artifacts and return files + warnings. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
  ): AssembleResult {
    const files: string[] = [];
    const warnings: string[] = [];
    const ctx = buildStackContext(config);
    this.generateCiWorkflow(
      outputDir, engine, ctx, files, warnings,
    );
    this.generateDockerfile(
      config, outputDir, resourcesDir, engine, ctx,
      files, warnings,
    );
    this.generateDockerCompose(
      config, outputDir, engine, ctx, files, warnings,
    );
    this.generateK8sManifests(
      config, outputDir, engine, ctx, files, warnings,
    );
    this.generateSmokeTestConfig(
      config, outputDir, resourcesDir, files, warnings,
    );
    this.generateDeployRunbook(
      outputDir, engine, ctx, files, warnings,
    );
    return { files, warnings };
  }

  /** CI workflow — always generated. */
  private generateCiWorkflow(
    outputDir: string,
    engine: TemplateEngine,
    ctx: Record<string, unknown>,
    files: string[],
    warnings: string[],
  ): void {
    const dest = path.join(
      outputDir, ".github", "workflows", "ci.yml",
    );
    if (renderAndWrite(engine, CI_TEMPLATE, dest, ctx)) {
      files.push(dest);
    } else {
      warnings.push("CI workflow template not found");
    }
  }

  /** Dockerfile — conditional on container === "docker". */
  private generateDockerfile(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
    ctx: Record<string, unknown>,
    files: string[],
    warnings: string[],
  ): void {
    if (config.infrastructure.container !== DOCKER_CONDITION) {
      warnings.push("Dockerfile skipped: container is not docker");
      return;
    }
    const stackKey = `${config.language.name}-${config.framework.buildTool}`;
    const templateFile = `dockerfile/Dockerfile.${stackKey}.njk`;
    const srcPath = path.join(
      resourcesDir, CICD_TEMPLATES, templateFile,
    );
    if (!fs.existsSync(srcPath)) {
      warnings.push(
        `Dockerfile template not found for stack: ${stackKey}`,
      );
      return;
    }
    const dest = path.join(outputDir, "Dockerfile");
    if (renderAndWrite(engine, templateFile, dest, ctx)) {
      files.push(dest);
    }
  }

  /** Docker Compose — conditional on container === "docker". */
  private generateDockerCompose(
    config: ProjectConfig,
    outputDir: string,
    engine: TemplateEngine,
    ctx: Record<string, unknown>,
    files: string[],
    warnings: string[],
  ): void {
    if (config.infrastructure.container !== DOCKER_CONDITION) {
      warnings.push(
        "Docker Compose skipped: container is not docker",
      );
      return;
    }
    const dest = path.join(outputDir, "docker-compose.yml");
    if (renderAndWrite(engine, COMPOSE_TEMPLATE, dest, ctx)) {
      files.push(dest);
    }
  }

  /** K8s manifests — conditional on orchestrator === "kubernetes". */
  private generateK8sManifests(
    config: ProjectConfig,
    outputDir: string,
    engine: TemplateEngine,
    ctx: Record<string, unknown>,
    files: string[],
    warnings: string[],
  ): void {
    if (config.infrastructure.orchestrator !== K8S_CONDITION) {
      warnings.push(
        "K8s manifests skipped: orchestrator is not kubernetes",
      );
      return;
    }
    for (const manifest of K8S_MANIFESTS) {
      const dest = path.join(outputDir, "k8s", manifest);
      const tpl = `k8s/${manifest.replace(".yaml", ".yaml.njk")}`;
      if (renderAndWrite(engine, tpl, dest, ctx)) {
        files.push(dest);
      }
    }
  }

  /** Smoke test config — conditional on smokeTests === true. */
  private generateSmokeTestConfig(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    files: string[],
    warnings: string[],
  ): void {
    if (!config.testing.smokeTests) {
      warnings.push(
        "Smoke test config skipped: smokeTests is false",
      );
      return;
    }
    const src = path.join(
      resourcesDir, CICD_TEMPLATES, SMOKE_SOURCE,
    );
    if (!fs.existsSync(src)) return;
    const dest = path.join(
      outputDir, "tests", "smoke", "smoke-config.md",
    );
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    fs.copyFileSync(src, dest);
    files.push(dest);
  }

  /** Deploy runbook — always generated. */
  private generateDeployRunbook(
    outputDir: string,
    engine: TemplateEngine,
    ctx: Record<string, unknown>,
    files: string[],
    warnings: string[],
  ): void {
    const dest = path.join(
      outputDir, "docs", "runbook", "deploy-runbook.md",
    );
    if (renderAndWrite(engine, RUNBOOK_TEMPLATE, dest, ctx)) {
      files.push(dest);
    } else {
      warnings.push("Deploy runbook template not found");
    }
  }
}

File: java/src/main/java/dev/iadev/assembler/CicdAssembler.java
package dev.iadev.assembler;

import dev.iadev.config.ContextBuilder;
import dev.iadev.domain.stack.LanguageCommandSet;
import dev.iadev.domain.stack.StackMapping;
import dev.iadev.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;

import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Coordinator that assembles CI/CD pipeline artifacts by
 * delegating to six specialized sub-assemblers.
 *
 * <p>This is the twenty-third assembler in the pipeline
 * (position 21 of 23 per RULE-005). It delegates to:
 * <ol>
 *   <li>{@link CiWorkflowAssembler} — always generated</li>
 *   <li>{@link CdWorkflowAssembler} — release always,
 *       deploy/rollback conditional on container</li>
 *   <li>{@link DockerfileAssembler} — conditional on
 *       {@code container == "docker"}</li>
 *   <li>{@link DockerComposeAssembler} — conditional on
 *       {@code container == "docker"}</li>
 *   <li>{@link K8sManifestAssembler} — conditional on
 *       {@code orchestrator == "kubernetes"}</li>
 *   <li>{@link SmokeTestAssembler} — conditional on
 *       {@code smokeTests == true}</li>
 * </ol>
 *
 * <p>Uses {@link #buildStackContext(ProjectConfig)} to resolve
 * stack-specific commands from {@link StackMapping}.</p>
 *
 * @see Assembler
 * @see StackMapping
 */
public final class CicdAssembler implements Assembler {

    private static final String CICD_TEMPLATES =
            "cicd-templates";

    /**
     * Lint command mapping per language-buildTool key.
     */
    static final Map<String, String> LINT_COMMANDS =
            Map.of(
                    "java-maven",
                    "./mvnw checkstyle:check",
                    "java-gradle",
                    "./gradlew spotlessCheck",
                    "kotlin-gradle",
                    "./gradlew ktlintCheck",
                    "typescript-npm",
                    "npm run lint",
                    "python-pip",
                    "ruff check .",
                    "go-go",
                    "golangci-lint run",
                    "go-go-mod",
                    "golangci-lint run",
                    "rust-cargo",
                    "cargo clippy -- -D warnings");

    private static final String DEFAULT_LINT_CMD =
            "echo 'No linter configured'";

    private static final int INITIAL_CICD_MAP_CAPACITY = 16;

    private final Path resourcesDir;
    private final CiWorkflowAssembler ciWorkflow;
    private final CdWorkflowAssembler cdWorkflow;
    private final DockerfileAssembler dockerfile;
    private final DockerComposeAssembler dockerCompose;
    private final K8sManifestAssembler k8sManifest;
    private final SmokeTestAssembler smokeTest;

    /**
     * Creates a CicdAssembler using classpath resources.
     */
    public CicdAssembler() {
        this(resolveClasspathResources());
    }

    /**
     * Creates a CicdAssembler with an explicit resources
     * directory.
     *
     * @param resourcesDir the base resources directory
     */
    public CicdAssembler(Path resourcesDir) {
        this.resourcesDir = resourcesDir;
        this.ciWorkflow = new CiWorkflowAssembler();
        this.cdWorkflow = new CdWorkflowAssembler();
        this.dockerfile = new DockerfileAssembler();
        this.dockerCompose = new DockerComposeAssembler();
        this.k8sManifest = new K8sManifestAssembler();
        this.smokeTest = new SmokeTestAssembler();
    }

    /**
     * {@inheritDoc}
     *
     * <p>Builds a stack context, then delegates to six
     * specialized sub-assemblers. Returns the merged list
     * of generated file paths.</p>
     */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        Map<String, Object> ctx =
                buildStackContext(config);
        Map<String, Object> fullContext =
                mergeContexts(
                        ContextBuilder.buildContext(config),
                        ctx);
        CicdContext cicdCtx = new CicdContext(
                config, outputDir, resourcesDir,
                engine, fullContext);

        CicdResult result = CicdResult.merge(List.of(
                ciWorkflow.assemble(cicdCtx),
                cdWorkflow.assemble(cicdCtx),
                dockerfile.assemble(cicdCtx),
                dockerCompose.assemble(cicdCtx),
                k8sManifest.assemble(cicdCtx),
                smokeTest.assemble(cicdCtx)));

        return result.files();
    }

    /**
     * Builds a stack-specific template context from
     * {@link StackMapping} constants.
     *
     * @param config the project configuration
     * @return the stack context map
     */
    static Map<String, Object> buildStackContext(
            ProjectConfig config) {
        String langKey = config.language().name()
                + "-" + config.framework().buildTool();

        Map<String, Object> ctx =
                new LinkedHashMap<>(
                        INITIAL_CICD_MAP_CAPACITY);
        addCommandEntries(langKey, ctx);
        addEnvironmentEntries(config, ctx);
        return ctx;
    }

    private static void addCommandEntries(
            String langKey,
            Map<String, Object> ctx) {
        LanguageCommandSet commands =
                StackMapping.LANGUAGE_COMMANDS.get(langKey);
        addBuildCommands(commands, ctx);
        addToolingCommands(langKey, commands, ctx);
    }

    private static void addBuildCommands(
            LanguageCommandSet commands,
            Map<String, Object> ctx) {
        ctx.put("compile_cmd",
                cmd(commands, LanguageCommandSet::compileCmd));
        ctx.put("build_cmd",
                cmd(commands, LanguageCommandSet::buildCmd));
        ctx.put("test_cmd",
                cmd(commands, LanguageCommandSet::testCmd));
        ctx.put("coverage_cmd",
                cmd(commands, LanguageCommandSet::coverageCmd));
    }

    private static void addToolingCommands(
            String langKey,
            LanguageCommandSet commands,
            Map<String, Object> ctx) {
        ctx.put("lint_cmd",
                LINT_COMMANDS.getOrDefault(
                        langKey, DEFAULT_LINT_CMD));
        ctx.put("file_extension",
                cmd(commands, LanguageCommandSet::fileExtension));
        ctx.put("build_file",
                cmd(commands, LanguageCommandSet::buildFile));
        ctx.put("package_manager",
                cmd(commands, LanguageCommandSet::packageManager));
    }

    private static String cmd(
            LanguageCommandSet commands,
            java.util.function.Function<
                    LanguageCommandSet, String> getter) {
        return commands != null ? getter.apply(commands) : "";
    }

    private static void addEnvironmentEntries(
            ProjectConfig config,
            Map<String, Object> ctx) {
        int port = StackMapping.FRAMEWORK_PORTS.getOrDefault(
                config.framework().name(),
                StackMapping.DEFAULT_PORT_FALLBACK);
        String healthPath =
                StackMapping.FRAMEWORK_HEALTH_PATHS
                        .getOrDefault(
                                config.framework().name(),
                                StackMapping
                                        .DEFAULT_HEALTH_PATH);
        String baseImage =
                StackMapping.DOCKER_BASE_IMAGES.getOrDefault(
                        config.language().name(),
                        StackMapping.DEFAULT_DOCKER_IMAGE);
        String resolvedImage = baseImage.replace(
                "{version}",
                config.language().version());

        ctx.put("framework_port", port);
        ctx.put("health_path", healthPath);
        ctx.put("docker_base_image", resolvedImage);
        ctx.put("container",
                config.infrastructure().container());
        ctx.put("native_build",
                config.framework().nativeBuild());
    }

    private static Map<String, Object> mergeContexts(
            Map<String, Object> base,
            Map<String, Object> override) {
        Map<String, Object> merged =
                new LinkedHashMap<>(base);
        merged.putAll(override);
        return merged;
    }

    private static Path resolveClasspathResources() {
        return dev.iadev.util.ResourceResolver
                .resolveResourcesRoot(CICD_TEMPLATES);
    }
}
